distributed/tests/test_worker.py

```diff
@@ -3805,3 +3805,57 @@ def test_unique_task_heap():
     assert heap.pop() == ts
 
     assert repr(heap) == "<UniqueTaskHeap: 0 items>"
+
+
+@gen_test(
+    # The timeout value here must be strictly larger than the default value for
+    # the timeout in `Worker.close`. The entire test should finish in <1s if
+    # everything works as intended but we want to make sure to not obfuscate
+    # another timeout therefore we increase this number.
+    timeout=60,
+)
+async def test_do_not_block_event_loop_during_shutdown():
+    async with Scheduler() as s:
+        async with Client(s.address, asynchronous=True) as c:
+            block_handler = asyncio.Event()
+            block_handler.clear()
+            called_handler = asyncio.Event()
+            called_handler.clear()
+            handler_continued = False
+
+            async def block_echo(x):
+                nonlocal handler_continued
+                called_handler.set()
+                await asyncio.sleep(0.1)
+                # If the event loop is blocked, we'll never hit this before the
+                # scheduler is closed
+                handler_continued = True
+                return x
+
+            s.handlers["block_echo"] = block_echo
+            w = await Worker(s.address)
+            addr = s.address
+
+            def f(x):
+                # This function calls into a handler on the remote which is
+                # blocked. This test eventually times out and the function
+                # should return and finish successfully, unblocking the
+                # ThreadPool
+                loop = get_worker().loop
+
+                async def _():
+                    pool = await ConnectionPool()
+                    scheduler_rpc = pool(addr)
+                    assert await scheduler_rpc.block_echo(x=x)
+
+                sync(loop, _)
```

Comment:
I think currently, the only way this test will fail is with a pytest-timeout at 300s. If you add a timeout here it should break the deadlock and fail the test with a TimeoutError

```suggestion
                sync(loop, _, callback_timeout="10s")
```

Reply:
I used your suggestion for a different unit test (I confirmed it reproduces the original issue)

FWIW, this test *did* fail but not after 300s but after 30s, i.e. after `Worker.close` / executor.shutdown reaches the timeout of 30s. The test would hit the `assert timeout_hit` and raise.

I prefer your test since it doesn't require a commpool and no `sync`